pkg/util/testing/client.go

```diff
@@ -18,22 +18,42 @@ package testing
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"strings"
 	"sync"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client/interceptor"
 
 	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
 	"sigs.k8s.io/kueue/pkg/controller/core/indexer"
 )
 
 func NewFakeClient(objs ...client.Object) client.Client {
-	return NewClientBuilder().WithObjects(objs...).WithStatusSubresource(objs...).Build()
+	return NewClientBuilder().WithObjects(objs...).WithStatusSubresource(objs...).
+		WithInterceptorFuncs(interceptor.Funcs{Patch: func(ctx context.Context, clnt client.WithWatch, obj client.Object, patch client.Patch, opts ...client.PatchOption) error {
+			// Apply patches are supposed to upsert, but fake client fails if the object doesn't exist,
+			// if an apply patch occurs for an object that doesn't yet exist, create it.
+			if patch.Type() != types.ApplyPatchType {
+				return clnt.Patch(ctx, obj, patch, opts...)
+			}
+			check, ok := obj.DeepCopyObject().(client.Object)
```

Comment:
I would expect you to override the type to `types.StrategicMergePatchType`.

Also, maybe we should have `NewFakeClientWithWeakSSA`, that we explicitly use when we are sure it's safe.

Reply:
The only issue is that we hit loads of these failures throughout Kueue.

Its whatever UpdateStatus is called so there are quite a few tests that fail with this.